### ISSUE TITLE
Add more math shorthands

### ISSUE THREAD

**Issue by mkorje** [typst/typst#4713]
### Description

With the current math shorthands, there are a number of other combinations that I'd "expect" to work. For example, `|-->` for `arrow.r.long.bar` and `<-|` for `arrow.l.bar`. I'd like to propose adding more shorthands, which I'd divide into two categories.

New shorthands that follow the established ones:
- `<-|` for `arrow.l.bar`
- `|-->` for `arrow.r.long.bar`
- `<--|` for `arrow.l.long.bar`
- `|->>` for `arrow.r.twohead.bar`
- `<<-|` for `arrow.l.twohead.bar`
- `|==>` for `arrow.r.double.long.bar`
- `<==|` for `arrow.l.double.long.bar`
- `<=|` for `arrow.l.double.bar`

New shorthands that extend the established ones:
- `->|` for `arrow.r.stop`
- `|<-` for `arrow.l.stop`
- `!->` for `arrow.r.not`
- `!<-` for `arrow.l.not`
- `!=>` for `arrow.r.double.not`
- `!<=` for `arrow.l.double.not`
- `!<->` for `arrow.l.r.not`
- `!<=>` for `arrow.l.r.double.not`
- `!<` for `lt.not`
- `!>` for `gt.not`
- `!<=` for `lt.eq.not`
- `!>=` for `gt.eq.not`

(I'll add more/change them if suggested)

Putting together a pull request for this is easy to do. But I was hoping to get some consensus on this. I'm also aware that we probably don't want to "bloat" the built-in shorthands.

EDIT: Adding suggestions from #2889.

Logic related.
- `|-` for `⊢` (`tack.r`)
- `-|` for `⊣` (`tack.l`)
- `|=` for `⊨` (`tack.r.double`)
- `=|` for `⫤` (`tack.l.double`)
- `!|-` for `⊬` (`tack.r.not`)
- `!|=` for `⊭` (`tack.r.double.not`)

Concatenated symbols.
- `-:` for `∹` (`dash.colon`)
- `!!` for `‼` (`excl.double`)
- `??` for `⁇` (`quest.double`)
- `?!` for `⁈` (`quest.excl`)
- `!?` for `⁉` (`excl.quest`)

### Use Case

For the first category, this would better align with what users would already expect to happen. Less so for the second category, though I think they would still make sense and be convenient.

**Comment by laurmaedje** [comment]
Prior discussion: https://github.com/typst/typst/issues/492

**Comment by laurmaedje** [comment]
I agree that the current shorthand list is not most consistent. However, I think an effort to improve the selection could then perhaps not try to fix this by exclusively adding new consistent pairings for the existing rare shorthands, but rather by also proposing removal of some existing rare ones.

cc @Enivex

**Comment by Enivex** [comment]
I think that rules around how shorthands can be used should be made stricter first. At a minimum they should be required to be surrounded by whitespace (or be alone).

**Comment by Enivex** [comment]
Another thing to consider is ligatures in popular coding fonts. Ideally they shouldn't conflict. Though I'm not sure that's an achievable goal.

**Comment by Enivex** [comment]
> but won't change anything inherently with shorthands

It will, because it'll prevent users from accidentally triggering shorthands (which can be incredibly annoying). Moreover, it'll allow more shorthands to be introduced without ambiguity.

**Comment by laurmaedje** [comment]
> Though I think we would need to implement a way for users to define their own shorthands before any removals happen to prevent breaking existing documents.

I don't think that's necessarily a blocker. It's relatively minor as far as breaking changes go and there is no stability guarantee.

**Comment by wrzian** [comment]
> I might add that I'd really like to see an ergonomic way to disable or enable individual shorthands.

Agreed. I think this is another desire that pushes towards runtime math parsing. My thoughts on how to do this ergonomically are currently jumbled, but I think this is quite possible.

**Comment by TimeTravelPenguin** [comment]
I just wanted to add that it appears that there is no `arrow.long.not` or `arrow.r.long.not`. For some reason, I thought it existed as `-/->` since there is already `-->`.
